pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -112,7 +109,18 @@ func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
 	if b.Spec.BackupConfig.S3BackupConfig == nil {
 		return b, nil
 	}
-	return b, c.deleteS3Snapshot(b)
+	// try to remove from s3 for 3 times, then give up. if we don't we get stuck forever
+	var delErr error
+	for i := 0; i < 3; i++ {
+		if delErr := c.deleteS3Snapshot(b); delErr != nil {
+			logrus.Warnf("failed to delete backup from s3: %v", delErr)
+		}
+		time.Sleep(2 * time.Second)
+	}
+	if delErr != nil {
+		logrus.Warnf("giving up on deleting backup [%s]: %v", b.Name, delErr)
+	}
+	return b, nil
```

Comment:
just want to understand why are we returning nil here? we should return err here so controller retries

Reply:
If I return err the controller will be suck forever tryin to delete a bad backup. I retry to delete and when I fail I just give up.

Reply:
This is fine for now. We follow the same pattern for the node provisioned by rancher. We retry removal on the remote provider several times, and then if it fails, we mark the node as removed anyway.